### PR TITLE
satellite/gracefulexit: handle piece not found messages from storagenode

### PR DESCRIPTION
**Pull request by mobyvb**
What: 
*  If a node claims to fail a transfer due to piece not found, remove that node from the pointer, delete the transfer queue item.
* If the pointer is piece hash verified, penalize the node. Otherwise, do not penalize the node.

Why:
A node might not have ever received a piece that ended up in an old, non-piece-hash-verified pointer. We do not want to unfairly penalize nodes for these pieces.
https://storjlabs.atlassian.net/browse/V3-3087

Please describe the tests:
 - Test 1: `TestFailureNotFoundPieceHashVerified` - Initiate graceful exit for a node holding one piece. The node responds to a transfer request with a transfer failed message and a `NOT_FOUND` error. Expect that the node is no longer in the pointer, and that the node has been marked as failing to transfer a piece.
 - Test 2: `TestFailureNotFoundPieceHashUnverified` - Initiate graceful exit for a node holding one piece. Update the pointer to have `PieceHashesVerified=false`. The node responds to a transfer request with a transfer failed message and a `NOT_FOUND` error. Expect that the node is no longer in the pointer, and that the node has *not* been marked as failing to transfer a piece.
 
Please describe the performance impact:

## Code Review Checklist (to be filled out by reviewer)
 - [x] NEW: Are there any Satellite database migrations? Are they forwards _and_ backwards compatible? 
 - [x] Does the PR describe what changes are being made?
 - [x] Does the PR describe why the changes are being made?
 - [x] Does the code follow [our style guide](https://github.com/storj/docs/blob/master/code/Style.md)?
 - [x] Does the code follow [our testing guide](https://github.com/storj/docs/blob/master/code/Testing.md)?
 - [x] Is the PR appropriately sized? (If it could be broken into smaller PRs it should be)
 - [x] Does the new code have enough tests? (*every* PR should have tests or justification otherwise. Bug-fix PRs especially)
 - [x] Does the new code have enough documentation that answers "how do I use it?" and "what does it do?"? (both source documentation and [higher level](https://github.com/storj/docs), diagrams?)
 - [x] Does any documentation need updating?
 - [x] Do the database access patterns make sense?

### CODE REVIEW

File: satellite/gracefulexit/endpoint.go [hunk 1]
```diff
@@ -755,7 +755,54 @@ func (endpoint *Endpoint) handleFailed(ctx context.Context, pending *pendingMap,
...
+		}
+
+		_, err = endpoint.metainfo.UpdatePieces(ctx, string(transfer.path), pointer, nil, []*pb.RemotePiece{nodePiece})
+		if err != nil {
+			return Error.Wrap(err)
+		}
+
+		// If the pointer was piece hash verified, we know this node definitely should have the piece
+		// Otherwise, no penalty.
+		if pointer.PieceHashesVerified {
```
Comment (VinozzZ) [inline satellite/gracefulexit/endpoint.go]: Maybe we can have a TODO in here to say that we should also not penalize nodes if the pointer has changed?

File: satellite/gracefulexit/endpoint.go [hunk 2]
```diff
@@ -755,7 +755,54 @@ func (endpoint *Endpoint) handleFailed(ctx context.Context, pending *pendingMap,
...
+	// If the pointer is not piece hash verified, do not count this as a failure.
+	if pb.TransferFailed_Error(errorCode) == pb.TransferFailed_NOT_FOUND {
+		endpoint.log.Debug("piece not found on node", zap.Stringer("node ID", nodeID), zap.ByteString("path", transfer.path), zap.Int32("piece num", transfer.pieceNum))
+		pointer, err := endpoint.metainfo.Get(ctx, string(transfer.path))
+		if err != nil {
+			return Error.Wrap(err)
+		}
+		remote := pointer.GetRemote()
+		if remote == nil {
+			return nil
```
Comment (VinozzZ) [inline satellite/gracefulexit/endpoint.go]: Should we delete the piece from the queue if there's no remote pieces?

File: satellite/gracefulexit/endpoint.go [hunk 3]
```diff
@@ -755,7 +755,54 @@ func (endpoint *Endpoint) handleFailed(ctx context.Context, pending *pendingMap,
...
+		pieces := remote.GetRemotePieces()
+
+		var nodePiece *pb.RemotePiece
+		for _, piece := range pieces {
+			if piece.NodeId == nodeID && piece.PieceNum == transfer.pieceNum {
+				nodePiece = piece
+			}
+		}
+		if nodePiece == nil {
+			return nil
```
Comment (VinozzZ) [inline satellite/gracefulexit/endpoint.go]: Same as above

File: satellite/gracefulexit/endpoint.go (hunk 1 above)
Comment (mobyvb) [reply satellite/gracefulexit/endpoint.go]: Shouldn't we do that as part of V3-3088? @ethanadams @VinozzZ

Comment (ethanadams) [reply satellite/gracefulexit/endpoint.go]: I agree.  It will get handled by V3-3088

**Pull request merged**